src/v/archival/ntp_archiver_service.cc

```diff
@@ -1367,4 +1398,172 @@ ntp_archiver::get_bucket_name() const {
     }
 }
 
+std::vector<std::reference_wrapper<housekeeping_job>>
+ntp_archiver::get_housekeeping_jobs() {
+    std::vector<std::reference_wrapper<housekeeping_job>> res;
+    if (_local_segment_merger) {
+        res.push_back(std::ref(*_local_segment_merger));
+    }
+    if (_remote_segment_merger) {
+        res.push_back(std::ref(*_remote_segment_merger));
+    }
+    return res;
+}
+
+ss::future<std::optional<upload_candidate_with_locks>>
+ntp_archiver::find_reupload_candidate(manifest_scanner_t scanner) {
+    if (!upload_loop_can_continue()) {
```

Comment:
It's fine if not, but I wonder if it'd be natural to reconcile/merge this and the upload logic with the existing code for periodic single-segment uploads.

Reply:
I decided to move it outside because of several things. First, there are jobs that can be done across the partition boundary (like deletion or scrapping). Second, the reupload from the cloud storage as a source is relatively expensive. One reupload job takes long time and holds two tcp connections from the pool. It takes long time because we need to download many segments one by one and each time we're accumulating latency. So, it makes sense to run only one such job at a time and this design achieves this. It'd be easier to do this inside the loop though.